### PR TITLE
Removed admin monitor related to ZFS

### PR DESCRIPTION
**Pull request by jglick**
This now just removes the `libzfs` library and the two things using it:

* `ZFSInstaller`, an admin monitor, likely not tested for years. If there is some point to this, better to resurrect as a plugin (and support on Linux ZFS too). AFAICT this was an old attempt by KK to improve the OOTB experience for novice admins installing Hudson on Solaris, that presumably does not match how Jenkins is managed today by the small number of Illumos users.
* ~`ZFSProvisioner`, literally unused and so also deleted in #5048.~

### Proposed changelog entries

* Removed administrative monitor offering to migrate `$JENKINS_HOME` on a ZFS filesystem.

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/java/hudson/lifecycle/SolarisSMFLifecycle.java [hunk 1]
```diff
@@ -1,57 +0,0 @@
...
-
-import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
-import jenkins.model.Jenkins;
-
-import java.io.IOException;
-import java.util.logging.Level;
-import java.util.logging.Logger;
-
-/**
- * {@link Lifecycle} for Hudson installed as SMF service.
```
Comment (jglick) [inline core/src/main/java/hudson/lifecycle/SolarisSMFLifecycle.java:34]: Used by https://github.com/jenkinsci/extras-ec2-launcher/blob/573b09eb9586e539dc30c8596e4886d167877b8e/src/main/resources/org/jvnet/hudson/ec2/launcher/smf.xml which I presume is also long obsolete.

**Comment by daniel-beck** [comment]
I think removing the ZFS specific stuff is reasonable, but unsure we should remove the `ProcessTree` support, that seems fairly self-contained?

It would be interesting to know how well this build still works on Solaris-based systems. @jimklimov ?

----

FWIW according to October usage stats, these are the popular OSes for nodes:

```
1058011 Nodes
 270091 N/A
 787920 Known OS, of which:
        552940 Linux (70.18%)
        198150 Windows (25.15%)
         30384 Mac OS X (3.86%)
          2606 SunOS (0.33%)
          2413 AIX (0.31%)
          1427 Other (0.18%), of which:
                    754 FreeBSD
                    402 HP-UX
                    192 z/OS
                     36 OS/400
                     20 OpenBSD
                     16 NONSTOP_KERNEL
                      6 NetBSD
                      1 DragonFlyBSD

**Comment by basil** [comment]
My company was a big illumos shop until [we began a migration to Linux in 2018](https://www.delphix.com/blog/kickoff-future-eko-2018), and in fact the Solaris 11 x86 support in `hudson.util.ProcessTree` was developed by a former Delphix employee in [JENKINS-40470](https://issues.jenkins.io/browse/JENKINS-40470) back in 2017. We never ran any controllers on illumos, but we used to run thousands of illumos agents. These days we only have a handful of legacy illumos agents remaining, as the company has almost entirely transitioned its development efforts to Linux. I have no personal affiliation with the illumos community, so I do not have a strong preference regarding this decision.

My gut feeling is that nobody would care too much if the fancy features like ZFS and SMF integration were removed, as I doubt those were used very much in the first place. I think most people who are running Jenkins controllers are doing so in Docker on Linux these days. But I think it would be a shame if the project dropped support for illumos agents. That functionality could be used by a project like AdoptOpenJDK, for example, to build OpenJDK on illumos, and it doesn't seem like a huge amount of effort to keep it in (basically just the `ProcessTree` stuff and the Sun PKCS11 workaround). I'm happy to try out any proposed change with an illumos x86 agent to verify that functionality keeps working.

**Comment by jimklimov** [comment]
@oleg-nenashev Thanks for the heads-up ;)

@all : I can't speak for the commercially branded Solaris, but the illumos community is quite active as far as project size and population allows, with several distributions actively developed. In particular, at least the OpenIndiana distribution happens on a Jenikns master with several rolling iterations daily, so might say Jenkins is frequently tested by that (haven't heard of issues arising in the past years).

Unfortunately, illumos bears the legacy of self-identifying as its close sibling `SunOS` although similar attempts to drop Solaris as irrelevant (luckily usual ending up in keeping compatibility for illumos and a better cooperation between the communities) led to an initiative to get another one defined. But you can guess that living with a name unrecognized by autotools, cmake, you name it, is even harder ;)

As for the changes in the PR themselves, I am not sure really what is or is not used in practice to make it all tick, or how to quickly and diligently find the corner cases to find out that something stopped ticking in a build with the code dropped. (And is also a matter of unavailable personal time due to too many backlogs already to catch up on, on my own side at least).

* The effort with LibZFS4J was reasonable on paper, and in fact is now far from a Solaris-like system feature nowadays, but the last time I looked (some years ago) Jenkins code only looked at it on "SunOS" systems. So to make better use of that, with snapshots at application-defined points where FS data across the wad of XML configs is consistent, and with clones and rollbacks and so on, and notably conflict-less backups (scraped from read-only snapshots) while Jenkins work is still happening in the live filesystem without a need to quietDown for longer than an FS flush of open buffers, the pendulum should go the other way around ;)

** Alas as it was, libzfs4j was more a nuisance because originally there was no public API/ABI in ZFS itself, so it was a hide-and-seek game about nailing the binary interface right for the private library for the system tools in the current host OS revision, or segfaulting with a wrong JNI signature otherwise. I think now there is an OpenZFS effort of a real library for applications which could make this effort more simple, although it could also be one of the ways that the same libzfs4j would try to get to managing the storage with its same Java API known by consumers.

** Another flawed point on low-speced systems was that Jenkins looked at all existing snapshots, not just those under its filesystem trees (JENKINS_HOME, workspaces, etc), and listed them in Java objects just because. Servers doing proper regular autosnaps to roll back in case of issues or user errors have a lot of snapshots. Heck, my laptop has 100-150k not-expired snapshots across several hundred datasets. It is cheap to make and store them, but is an expensive PITA to read them all on slower storage, and an unbearable weight to cache them into Java lists (Jenkins OOM'ed on as many gigs of `-Xmx` as I could throw at it).

* For SMF integration not sure, there is a manifest in OpenIndiana for example, at https://github.com/OpenIndiana/oi-userland/blob/oi/hipster/components/developer/jenkins/files/jenkins-standalone.xml - looking at code being dropped, seems there was recognition of running as a service for proper activities to restart (die and let the system bring Jenkins back up) etc.

* No idea about the `/proc/PID/*` integration bits, and how life would change without them (it is more a matter of why and how Jenkins uses that - if for proper tracking of child processes in builds, then probably a must-have)

* Elevation with RBAC-aware `pfexec` is the right way, compared to simple `sudo` et al, though not sure if the code dropped in SU.java allowed to select one or another. Tools like `su` and `sudo` are often used, especially where oldschool people re-enabled the `root` account. By default `root` does not exist as an interactive user (except in single-user repair boots), it is a role that an application context can assume if allowed - in that case sudo et al are worthless.

* No opinion on PKCS

**Comment by jimklimov** [comment]
By the way, looking at the stats above, I *think* the numbers for BSD in particular may be misleading. As a project with decades of patch reviews over e-mail and everybody testing new builds on their rigs with different archtectures, I believe they spent several recent years going into modern ways of production - including automation of builds, tests and eventually CI. So *them* not being large Jenkins users today might mean just that they are starting to be, would be interesting to see if their numbers rise "tomorrow".

**Comment by daniel-beck** [comment]
Different conversation, but you can see the stats at http://stats.jenkins.io/jenkins-stats/svg/svgs.html for every month for the last decade or so (column "nodes") if you want to know more.

**Comment by jimklimov** [comment]
Thanks @daniel-beck 

Overall, don't merge this just yet - I'll stir up the question in illumos discussions so maybe people better aware about the lower-level details, and whether some of the proposed removals impact running Jenkins on illumos, could chime in here.

It seems that one large part of this PR, at least by file count, is the ZFS integration in the installer and wizard and potentially other parts of code. This one I think (as a libzfs4j co-maintainer) is safe to excise at this point. That said, the "design benefits" from having such integration can exist, and I hope some day someone brings it back, revised and extended for supporting the modern branches (ZoL, ZoB, ZoI, ZoM...) and doing that not so resource-hungry and crash-prone manner.

Given that for reasonable snapshots to happen, care must be taken to flush all relevant files (quietDown during backup is needed AFAIK to avoid ZIPing half-written XMLs or files referring to different states of the overall data tree by the time zipper gets to this or that file individually... something that freeze+flush+snap+unfreeze solves easily), I am not sure such comeback would be possible as a plugin however.

**Comment by oleg-nenashev** [comment]
One of the options would be to detach all the code to a plugin, but without a "detach" metadata in the Jenkins core.After this change any Solaris/Illumos/OpenIndiana user will need to install the plugin to get the support working. It is better than just removing the code, though it does not seem that all the removed code is covered by extension points

File: core/src/main/java/hudson/os/solaris/ZFSProvisioner.java [hunk 2]
```diff
@@ -1,137 +0,0 @@
...
-import org.jenkinsci.Symbol;
-import org.jvnet.solaris.libzfs.LibZFS;
-import org.jvnet.solaris.libzfs.ZFSFileSystem;
-
-/**
- * {@link FileSystemProvisioner} for ZFS.
- *
- * @author Kohsuke Kawaguchi
- */
-public class ZFSProvisioner extends FileSystemProvisioner implements Serializable {
```
Comment (jglick) [inline core/src/main/java/hudson/os/solaris/ZFSProvisioner.java:51]: Note that as in #5048, this is not just unused but unusable.

**Comment by jglick** [comment]
> the OpenIndiana distribution happens on a Jenikns master

Please be specific about whether the ~master~ _controller_ or the _agent_ is running a Solaris flavor. Matters a great deal when discussing particular parts of the code.

**Comment by jimklimov** [comment]
Yeah, sorry about the naming, everyone invents their own so easy to get lost.

Looking at recent builds like https://hipster.openindiana.org/jenkins/job/oi-userland/5567/consoleFull it seems the controller is running the builds locally, without an agent. IIRC lately as part of the build job it might spin up a local zone (container) to run a clean build in, but that is still driven from the same job's Makefile so agent-less.

I was not directly involved in the infra, so can't say much more without considerable guesswork.

**Comment by jglick** [comment]
> the controller is running the builds locally, without an agent

Right, so it would make sense to make sure that Jenkins can run both controller and agent without error, at least. Whether the lifecycle is important, I am not sure. Most people these days prefer to run in containers which thus do not need custom lifecycle implementations (restarting Jenkins means killing and recreating the container).

**Comment by jvz** [comment]
What if you want Jenkins to control the containers like for docker agents and such? I've wondered how you could structure Jenkins to use jails and snapshotting to get a bunch of nifty security and ops features for free, though I haven't tried it yet. ZFS-like filesystems are also developed for other operating systems, so the idea sounds pretty neat.

**Comment by basil** [comment]
The hardware-accelerated cryptographic AES implementation in the Sun PKCS11 provider was still buggy as of 2018, so it needs to be disabled when running on illumos. However I do not feel strongly that this must be done with Java code in Jenkins core itself. The user can do this by setting the `-Dsun.security.pkcs11.enable-solaris=false` JVM option or removing the Sun PKCS11 provider from the `java.security` configuration file. I think it would be acceptable to remove the relevant Java code from Jenkins core so long as the documentation and/or upgrade guide was updated to reflect this.

**Comment by jclulow** [comment]
I am a member of the @illumos core team, and I know there are several projects in our community that run both controllers and agents directly on illumos systems; e.g., the OpenIndiana and the Joyent/SmartOS folks.  Jenkins, alongside Gerrit and OpenGrok, are important applications for our community and amongst the key reasons for our continued interest in supporting new JDKs on our platform (as per work that @ptribble and others are doing).

> The hardware-accelerated cryptographic AES implementation in the Sun PKCS11 provider was still buggy as of 2018, so it needs to be disabled when running on illumos.

I'm not aware of the specific bugs you're talking about (do you have any references to the illumos bug tracker for these issues?), but I'm also not sure if this is the case anymore.  I believe @jasonbking has done some work in the last two years on PKCS11 in illumos; perhaps he can comment.  Either way I agree that it seems like a workaround that Jenkins ought not to have to carry directly.

I took a look at this change, and I think there are some parts that seem fine to throw out.  In particular, `libzfs.so.1` isn't a stable/public interface, so depending on it from code outside of the core of illumos is likely to continue to create compatibility issues over time.  It isn't completely clear to me that either the "ZFSInstaller" or "ZFSProvisioner" bits are terribly useful -- I expect those deploying Jenkins will make their own file system layout choices on their own.  Presumably all of that stuff can be removed without presenting and serious challenges.

Things that are likely to be in use, though, and which it would be good to keep include:
* support for `pfexec` as a `sudo`-like mechanism for user switching
* the basic awareness of [smf(5)](https://illumos.org/man/5/smf) as a service supervisor that appears to exist today

We in illumos are a small but active community, and we do variously make use of Jenkins for our build infrastructure.  Let us know how we can help keep the core support for illumos around in Jenkins!

**Comment by jasonbking** [comment]
The PKCS11 issue is likely [illumos#4912](https://www.illumos.org/issues/4912). While it's on my list of PKCS11 issues to fix, I haven't made it to tackling that (yet).

**Comment by 3eka** [comment]
I guess this is used on Illumos based OS: https://hipster.openindiana.org/jenkins/
Not sure if your change will break it... Is it so urgent and code is so big that you need to remove it?

Regards.

**Comment by jperkin** [comment]
The Jenkins controller that handles the pkgsrc package build jobs for SmartOS/illumos/macOS/Linux/NetBSD from https://pkgsrc.joyent.com runs on illumos.  Please don't remove it.

**Comment by jglick** [comment]
@jimklimov @jclulow when you get a chance, could you take a look to make sure this is not removing anything actually used by anyone?

**Comment by jimklimov** [comment]
Sorry, had a hectic few weeks lately; is there a WAR file to fetch and be sure I'm testing the same thing you intend to get after PR merge? :)

**Comment by jimklimov** [comment]
On a side note, probably this change would make it easier to build Jenkins plugins on illumos systems - those which require a version of Jenkins core tend to have it segfault during test startup, unless I took care to export the "lockpicked" envvar settings for libzfs4j ABI to use, and in turn that only helps if the required version of Jenkins core is new enough to include the libzfs4j new enough to know about these tweaks.

Not that the problem would go away overnight after merge, but as dependencies for plugins would get updated to one where this drop is merged, building against them "out of the box" would become more simple.

**Comment by jclulow** [comment]
Sorry for the delay.  I think this removes the libzfs parts that have always been a bit of a problem, without cutting out anything else that might be helpful!  Thanks for revising the plan here!

**Comment by jglick** [comment]
> is there a WAR file to fetch and be sure I'm testing the same thing you intend to get after PR merge?

If and when c874764 passes all tests, https://repo.jenkins-ci.org/incrementals/org/jenkins-ci/main/jenkins-war/2.269-rc30602.c8747645361a/jenkins-war-2.269-rc30602.c8747645361a.war should be a valid link.

**Comment by jimklimov** [comment]
Thanks for the patience, finally got to test this in practice. Deployed on a recent OpenIndiana Hipster 2020.10 and ran with `JENKINS_HOME=/var/tmp/j269 java -jar jenkins-war-2.269-rc30602.c8747645361a.war` and it started out of the box, no fiddling with libzfs4j ABI anymore ;)

Got some plugins, defined jobs and ran a build on master, seems to work okay. Aborting a running job by red cross also works :)

**Comment by jglick** [comment]
> just started out of the box, no fiddling with libzfs4j ABI anymore

Good!

```
java.io.IOException: Failed to exec '' No such file or directory
        at hudson.lifecycle.UnixLifecycle.restart(UnixLifecycle.java:87)
```

Probably the native interfaces used for exec-to-self do not work on Solaris. Frankly I do not think we should default to using `UnixLifecycle`, or should at least rename in `LinuxLifecycle` and activate it only on Linux. Proper server administration on a VM means using an initialization service, or on a container system means leaving restarts to the orchestrator, so IMO `-Dhudson.lifecycle=hudson.lifecycle.ExitLifecycle` would be a saner default these days (and it is pure Java too). There is a `SolarisLifecycle` that gets selected if `SMF_FMRI` or `SMF_RESTARTER` are set, whatever those are; it is basically `ExitLifecycle` but with a zero exit code.

At any rate, clearly unrelated to this change.

**Review by oleg-nenashev — APPROVED** [review]
Taking the feedback from @jimklimov and @jclulow, I think we are good to go here.

**Review by res0nance — APPROVED** [review]

**Review by jimklimov — APPROVED** [review]

**Comment by jimklimov** [comment]
The (non-service run, just from terminal) restart issue is intermittent. Sometimes (rather often) the checkbox to restart Jenkins after applying new plugins just works in this test env too :)

**Comment by jimklimov** [comment]
By the way, if someone later undertakes reintroduction of similar ZFS awareness as a plugin, is there some doc on what are the steps to grab dropped core code and paste it into a plugin boilerplate as first iteration?

**Comment by jglick** [comment]
> reintroduction of similar ZFS awareness

I would hope via forked processes rather than JNI/JNA…

> plugin boilerplate

https://www.jenkins.io/doc/developer/tutorial/create/ etc.

**Pull request merged**